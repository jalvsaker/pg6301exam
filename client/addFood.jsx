import { useState } from "react";

export function AddFood({ reload }) {
  const [name, setName] = useState("");
  const [price, setPrice] = useState("");

  async function handleSubmit(e) {
    e.preventDefault();

    const body = { name, price };

    await fetch("/api/foods", {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify(body),
    });

    reload();
  }

  return (
    <form onSubmit={handleSubmit}>
      <label>
        Name:
        <input
          type={"text"}
          value={name}
          onChange={(e) => setName(e.target.value)}
        />
      </label>
      <label>
        Price:
        <input
          type={"number"}
          value={price}
          onChange={(e) => setPrice(e.target.value)}
        />
      </label>
      <button>Submit</button>
    </form>
  );
}
