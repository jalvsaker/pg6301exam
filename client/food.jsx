export function Food({ food }) {
  return (
    <div>
      {food.name} - {food.price}
      <div>
        <input type="number" /> <button>Add to cart</button>
      </div>
      <hr />
    </div>
  );
}
