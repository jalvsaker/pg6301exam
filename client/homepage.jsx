export function Homepage() {
  return <h1>Hello world</h1>;
}
