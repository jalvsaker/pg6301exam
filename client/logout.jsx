import { useNavigate } from "react-router-dom";
import { useEffect } from "react";

export function Logout({ setUser }) {
  const navigate = useNavigate();

  useEffect(() => {
    (async () => {
      const res = await fetch("/api/login", { method: "delete" });
      if (res.ok) {
        setUser({});
        navigate("/");
      }
    })();
  }, []);

  return "Logging out...";
}