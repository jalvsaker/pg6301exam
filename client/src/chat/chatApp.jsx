import { useState } from "react";

export function ChatApp({ messages, onNewMessage }) {
  const [message, setMessage] = useState("");

  function handleSubmit(e) {
    e.preventDefault();
    onNewMessage(message);
    setMessage("");
  }

  return (
    <>
      <h1>Chat</h1>
      <hr />
      {messages.map(({ message, username }, index) => (
        <div key={index}>
          <strong>{username}</strong> - {message}
          <hr />
        </div>
      ))}
      <form onSubmit={handleSubmit}>
        <input
          type="text"
          value={message}
          onChange={(e) => setMessage(e.target.value)}
          autoFocus={true}
        />
        <button>Send</button>
      </form>
    </>
  );
}
