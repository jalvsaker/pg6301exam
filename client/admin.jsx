import { useNavigate } from "react-router-dom";
import { useEffect, useState } from "react";
import { AddFood } from "./addFood";

export function Admin({ user }) {
  const navigate = useNavigate();

  if (!user.isAdmin) {
    navigate("/");
  }

  const [foods, setFoods] = useState([]);
  const [loading, setLoading] = useState(true);
  const [reloader, setReloader] = useState(false);

  useEffect(() => {
    (async () => {
      const res = await fetch("/api/foods");
      if (res.ok) {
        setFoods(await res.json());
        setLoading(false);
      }
    })();
  }, [reloader]);

  function reload() {
    setReloader((r) => !r);
  }

  if (loading) {
    return <div>Loading...</div>;
  }

  return (
    <>
      <h1>Admin panel</h1>

      <AddFood reload={reload} />

      {foods.map((food) => (
        <div key={food._id}>{food.name}</div>
      ))}
    </>
  );
}
