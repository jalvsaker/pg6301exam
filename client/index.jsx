import React from "react";
import { createRoot } from "react-dom/client";
import {
  BrowserRouter,
  Navigate,
  NavLink,
  Route,
  Routes,
} from "react-router-dom";
import { Homepage } from "./homepage";

function App() {
  return (
    <BrowserRouter>
      <nav>
        <NavLink to={"/"}>Home</NavLink>
        <NavLink to={"/test"}>Test</NavLink>
      </nav>
      <Routes>
        <Route path={"/"} element={<Homepage />} />
        <Route path={"/test"} element={<h2>test</h2>} />
        <Route path={"/*"} element={<Navigate to={"/"} />} />
      </Routes>
    </BrowserRouter>
  );
}

const root = createRoot(document.getElementById("root"));
root.render(
  <React.StrictMode>
    <App />
  </React.StrictMode>
);
