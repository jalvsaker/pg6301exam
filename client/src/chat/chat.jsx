import { useEffect, useState } from "react";
import { ChatApp } from "./chatApp";

export function Chat({ user }) {
  const [messages, setMessages] = useState([]);
  const [ws, setWs] = useState();

  useEffect(() => {
    const ws = new WebSocket(window.location.origin.replace(/^http/, "ws"));
    ws.onmessage = (event) => {
      const { username, message } = JSON.parse(event.data);
      setMessages((messages) => [...messages, { username, message }]);
    };
    setWs(ws);
  }, []);

  function handleNewMessage(message) {
    ws.send(JSON.stringify({ message, username: user.username }));
  }

  return <ChatApp messages={messages} onNewMessage={handleNewMessage} />;
}
