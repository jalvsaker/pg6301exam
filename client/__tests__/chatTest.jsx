import React from "react";
import { createRoot } from "react-dom/client";
import { act, Simulate } from "react-dom/test-utils";
import { ChatApp } from "../src/chat/chatApp";
import { Chat } from "../src/chat/chat";


describe("chat tests", function () {
  it("should submit chat", async () => {
    const element = document.createElement("div");
    const root = createRoot(element);

    const mock = jest.fn();

    await act(() => {
      root.render(
        <ChatApp
          messages={[
            { username: "user 1", message: "Hello" },
            { username: "user 2", message: "Hey" },
          ]}
          onNewMessage={mock}
        />
      );
    });

    await act(() => {
      Simulate.change(element.querySelector("input"), {
        target: { value: "hei" },
      });
    });

    await act(() => {
      Simulate.submit(element.querySelector("form"));
    });

    expect(element.innerHTML).toMatchSnapshot();
    expect(mock).toBeCalledWith("hei");
  });

  it("should show chat", async () => {
    const element = document.createElement("div");
    const root = createRoot(element);

    await act(() => {
      root.render(
        <Chat/>
      );
    });

    expect(element.innerHTML).toMatchSnapshot();
  });
});
