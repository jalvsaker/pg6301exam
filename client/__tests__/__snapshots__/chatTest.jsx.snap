// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`chat tests should submit chat 1`] = `"<h1>Chat</h1><hr><div><strong>user 1</strong> - Hello<hr></div><div><strong>user 2</strong> - Hey<hr></div><form><input type="text" value=""><button>Send</button></form>"`;
